server/sonar-server/src/main/java/org/sonar/server/rule/CachingRuleFinder.java

```diff
@@ -0,0 +1,196 @@
+  }
+
+  @Override
+  @CheckForNull
+  public Rule find(@Nullable RuleQuery query) {
+    if (query == null) {
+      return null;
+    }
+
+    return rulesByRuleDefinition.entrySet().stream()
+      .filter(entry -> matchQuery(entry.getKey(), query))
+      .sorted(FIND_BY_QUERY_ORDER)
+      .map(Map.Entry::getValue)
+      .findFirst()
+      .orElse(null);
+  }
+
+  @Override
+  public Collection<Rule> findAll(@Nullable RuleQuery query) {
+    if (query == null) {
+      return Collections.emptyList();
+    }
+    return rulesByRuleDefinition.entrySet().stream()
+      .filter(entry -> matchQuery(entry.getKey(), query))
+      .sorted(FIND_BY_QUERY_ORDER)
+      .map(Map.Entry::getValue)
+      .collect(MoreCollectors.toList());
+  }
+
+  private static boolean matchQuery(RuleDefinitionDto ruleDefinition, RuleQuery ruleQuery) {
+    if (RuleStatus.REMOVED.equals(ruleDefinition.getStatus())) {
+      return false;
+    }
+    String repositoryKey = ruleQuery.getRepositoryKey();
+    if (ruleQuery.getRepositoryKey() != null && !repositoryKey.equals(ruleDefinition.getRepositoryKey())) {
+      return false;
+    }
+    String key = ruleQuery.getKey();
+    if (key != null && !key.equals(ruleDefinition.getRuleKey())) {
+      return false;
+    }
+    String configKey = ruleQuery.getConfigKey();
+    return configKey == null || configKey.equals(ruleDefinition.getConfigKey());
+  }
+
+  private static Rule toRule(RuleDefinitionDto ruleDefinition, List<RuleParamDto> params) {
+    String severity = ruleDefinition.getSeverityString();
+    String description = ruleDefinition.getDescription();
+    RuleDto.Format descriptionFormat = ruleDefinition.getDescriptionFormat();
+
+    org.sonar.api.rules.Rule apiRule = new org.sonar.api.rules.Rule();
+    apiRule
+      .setName(ruleDefinition.getName())
+      .setLanguage(ruleDefinition.getLanguage())
+      .setKey(ruleDefinition.getRuleKey())
+      .setConfigKey(ruleDefinition.getConfigKey())
+      .setIsTemplate(ruleDefinition.isTemplate())
+      .setCreatedAt(new Date(ruleDefinition.getCreatedAt()))
+      .setUpdatedAt(new Date(ruleDefinition.getUpdatedAt()))
+      .setRepositoryKey(ruleDefinition.getRepositoryKey())
+      .setSeverity(severity != null ? RulePriority.valueOf(severity) : null)
+      .setStatus(ruleDefinition.getStatus().name())
+      .setTags(ruleDefinition.getSystemTags().toArray(new String[ruleDefinition.getSystemTags().size()]))
+      .setId(ruleDefinition.getId());
+    if (description != null && descriptionFormat != null) {
+      if (RuleDto.Format.HTML.equals(descriptionFormat)) {
+        apiRule.setDescription(description);
+      } else {
+        apiRule.setDescription(Markdown.convertToHtml(description));
+      }
+    }
+
+    List<org.sonar.api.rules.RuleParam> apiParams = newArrayList();
+    for (RuleParamDto param : params) {
+      apiParams.add(new org.sonar.api.rules.RuleParam(apiRule, param.getName(), param.getDescription(), param.getType())
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "RuleParam"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)